PWGCF/GenericFramework/Tasks/flowGenericFramework.cxx

```diff
@@ -278,163 +320,188 @@ struct GenericFramework {
+    // auto multT0C = collision.multFT0C();
+    auto multNTracksPV = collision.multNTracksPV();
+
+    if (vtxz > vtxZup || vtxz < vtxZlow)
+      return 0;
+    if (multNTracksPV < fMultPVCutLow->Eval(centrality))
+      return 0;
+    if (multNTracksPV > fMultPVCutHigh->Eval(centrality))
+      return 0;
+    if (multTrk < fMultCutLow->Eval(centrality))
+      return 0;
+    if (multTrk > fMultCutHigh->Eval(centrality))
+      return 0;
+    if (multTrk > fMultMultPVCut->Eval(multNTracksPV))
+      return 0;
+
+    return 1;
+  }
 
   enum datatype {
-    kRaw,
-    kDerived
+    kData,
+    kDerivedData,
+    kGen
   };
 
   template <datatype dt, typename TCollision, typename TTracks>
-  void processData(TCollision collision, TTracks tracks, float centrality)
+  void processCollision(TCollision collision, TTracks tracks, const float& centrality)
   {
     if (tracks.size() < 1)
       return;
     if (centrality < centbinning.front() || centrality > centbinning.back())
       return;
     float vtxz = collision.posZ();
-
-    if (cfgFillQA) {
-      QAregistry.fill(HIST("cent_nch"), tracks.size(), centrality);
-      QAregistry.fill(HIST("vtxZ"), vtxz);
-    }
+    if (cfgFillQA)
+      (dt == kGen) ? QAregistry.fill(HIST("vtxZ_gen"), vtxz) : QAregistry.fill(HIST("vtxZ"), vtxz);
     fGFW->Clear();
     float l_Random = fRndm->Rndm();
-    float weff = 1, wacc = 1;
-
     for (auto& track : tracks) {
-      if (cfgFillWeights)
-        fWeights->Fill(track.phi(), track.eta(), vtxz, track.pt(), centrality, 0);
-      if (cfg.mEfficiency)
-        weff = cfg.mEfficiency->GetBinContent(cfg.mEfficiency->FindBin(track.pt()));
-      else
-        weff = 1.0;
-      if (weff == 0)
-        continue;
-      weff = 1. / weff;
-      if (cfg.mAcceptance)
-        wacc = cfg.mAcceptance->GetNUA(track.phi(), track.eta(), vtxz);
-      else
-        wacc = 1;
-      registry.fill(HIST("phi_eta_vtxZ_corrected"), track.phi(), track.eta(), vtxz, wacc);
-      bool WithinPtPOI = (ptpoilow < track.pt()) && (track.pt() < ptpoiup); // within POI pT range
-      bool WithinPtRef = (ptreflow < track.pt()) && (track.pt() < ptrefup); // within RF pT range
-      if (WithinPtRef)
-        fGFW->Fill(track.eta(), fPtAxis->FindBin(track.pt() - 1), track.phi(), wacc * weff, 1);
-      if (WithinPtPOI)
-        fGFW->Fill(track.eta(), fPtAxis->FindBin(track.pt()) - 1, track.phi(), wacc * weff, 2);
-      if (WithinPtPOI && WithinPtRef)
-        fGFW->Fill(track.eta(), fPtAxis->FindBin(track.pt()) - 1, track.phi(), wacc * weff, 4);
+      ProcessTrack(track, centrality, vtxz);
+    }
+    for (uint l_ind = 0; l_ind < corrconfigs.size(); l_ind++) {
+      FillFC((dt == kGen) ? fFC_gen : fFC, corrconfigs.at(l_ind), (cfgUseNch) ? tracks.size() : centrality, l_Random);
+    }
+  }
 
+  template <typename TrackObject>
+  inline void ProcessTrack(TrackObject const& track, const float& centrality, const float& vtxz)
+  {
+    float weff = 1, wacc = 1;
+    float pt, eta, phi;
```

Comment:
For the next iteration consider not using these global variables
Use the direct expressions instead (`track.pt()` or `mcParticle.eta()` for instance) and let the compiler optimize on how to allocate them